Add flexible rpc client to tendermint

ethanfrey opened the pull request:
This code to call the rpc was in light-client, but I abstracted it to make it more testable, and provided mocks, so we can actually can simulate the server. 

A bit of clean up of the existing code to test the rpc interface, and then a lot of code in `rpc/client/*`

I would like to get this in by 0.9.0 and then use it as the basis for light-client (so I can clean up that code as well).

Please let me know if I should change anything.  Mocks are not finished, especially as faking sigs is hard, but they are halfway done and already provide a boon for testing (`client/helpers_test.go`).

ethanfrey commented:
Forgot to mention, this is in response to issue #416

melekes commented:
Wow! So much code!

melekes commented on rpc/client/local/app_test.go:
```diff
@@ -0,0 +1,45 @@
+)
+
+// MakeTxKV returns a text transaction, allong with expected key, value pair
+func MakeTxKV() ([]byte, []byte, []byte) {
```
not a fan of go-common, but code duplication here hurts me a bit.

ethanfrey replied:
Didn't know this was in go-common, will take a look and fix

melekes replied:
No, I meant that we should probably move this into one package and reuse rather than duplicating the code.

ethanfrey replied:
that code belongs in dummy actually... will put it there.

ethanfrey replied:
RandAsciiBytes is almost the same as common.RandStr, will use RandStr instead

melekes commented on rpc/client/local/client.go:
```diff
@@ -0,0 +1,139 @@
+package local
+
+import (
+	nm "github.com/tendermint/tendermint/node"
```
nm - node module?

ethanfrey replied:
yup, i use node too much as a local variable and didn't want naming conflict

melekes commented on rpc/client/local/rpc_test.go:
```diff
@@ -0,0 +1,183 @@
+package local_test
```
also, this testing suite is duplicated in 2 places. I understand we want to test both implementations. Not sure what is the best way to deal with this.

ethanfrey replied:
Yes, this hurts me a bit too. If you have a clean solution so I can convert this to a table-driven test, let me know.

Also, for code coverage, it only really measures the coverage from tests in the same package.

I'll think of a way to do code reuse. Let me know if it is too ugly.

melekes commented on rpc/client/mock/abci.go:
```diff
@@ -0,0 +1,186 @@
+	App abci.Application
+}
+
+func (a ABCIApp) _assertABCIClient() client.ABCIClient {
```
Why do we need such methods (`_assertIsClient`, `_assertABCI...`)?

ethanfrey replied:
Not needed by nice helpers. Talked with Bucky about them, he liked the idea too.

Basically, when you create Struct explicitly to fulfill an interface, you can assert so if the methods don't match, you get a compile error in the module itself, not when someone else tried to use it.

Matt had this problem with endblock changing SIG and then quietly being ignored in basecoin without any compiler warnings.

ebuchman replied:
we don't need this anymore. It was only an issue because there was an embedded interface type in abci.Application (the BlockchainAware bit). Now that has been moved into the abci.Application, so as long as that interface is being passed around (it is, everywhere), then the compiler already type checks

ethanfrey replied:
In this case, I want to make sure my mock structs properly implement the client interface. 

the abci app was just to explain to anton why the idea came up.

melekes commented on rpc/client/mock/abci.go:
```diff
@@ -0,0 +1,186 @@
+}
+
+func (a ABCIApp) BroadcastTxAsync(tx types.Tx) (*ctypes.ResultBroadcastTx, error) {
+	d := a.App.DeliverTx(tx)
```
I think it should be `CheckTx` here, no? and empty result in this case

```
// Returns right away, with no response
func BroadcastTxAsync(tx types.Tx) (*ctypes.ResultBroadcastTx, error) {
	err := mempool.CheckTx(tx, nil)
	if err != nil {
		return nil, fmt.Errorf("Error broadcasting transaction: %v", err)
	}
	return &ctypes.ResultBroadcastTx{}, nil
}
```

ethanfrey replied:
Yes, will fix. Good eyes.

ethanfrey replied:
We want this action to actually happen in the mock, but I change it to return the results of CheckTx, and trigger DeliverTx in a go-routine on success, so we can actually simulate this race in test cases if we want.

melekes commented on rpc/client/mock/abci.go:
```diff
@@ -0,0 +1,186 @@
+}
+
+func (a ABCIApp) BroadcastTxSync(tx types.Tx) (*ctypes.ResultBroadcastTx, error) {
+	d := a.App.DeliverTx(tx)
```
same here (`CheckTx`)

melekes commented on rpc/test/helpers.go:
```diff
@@ -1,98 +1,135 @@
-	clientJSON = client.NewClientJSONRPC(requestAddr)
-	clientGRPC = core_grpc.StartGRPCClient(grpcAddr)
+// f**ing long, but unique for each test
+func makePathname() string {
```
https://golang.org/pkg/io/ioutil/#TempDir ?

melekes commented:
![](http://www.reactiongifs.com/r/zgtu.gif)

Nice job! Learned a lot even by reading the code!

ethanfrey commented on rpc/client/http/rpc_test.go:
```diff
@@ -0,0 +1,211 @@
+
+	// subscribe to a transaction event
+	_, _, tx := MakeTxKV()
+	// this causes a panic in tendermint core!!!
```
old comment... I think I pushed a fix to tendermint in October, before I got hired even.

this client code was born out of a personal project.  But yeah, due to some typos and crazy interface magic, subscribing to a tx event would cause tendermint to panic:
https://github.com/tendermint/tendermint/issues/285

Removing this comment now.

ethanfrey commented:
@melekes I pulled local and http subpackages into client, renamed all structs, rewrote the rpc_tests to be table driven and accept any number of client.Client implementations.

Oh, and rebased everything on develop (which is why all commits are "new")

Please take a look at 7fe9ff5141457e3cc05ebc87bfb7f561c28e52c8 and let me know what you think.

ethanfrey commented:
@ebuchman I did lots of cleanup from anton's comments, please take a look.  I'd like to pull out the code in light-client.

melekes commented:
LGTM

ebuchman commented on rpc/client/rpc_test.go:
```diff
@@ -0,0 +1,236 @@
+	select {
+	case res := <-r:
+		checkData(res, ctypes.ResultTypeEvent)
+		// this is good.. let's get the data... ugh...
```
hm

ethanfrey commented:
I'll clean this up tomorrow. Thanks for the review.

Yeah, all the event stuff issue stuff is not really supported well.  I added what I had to the client.HTTP only and the user has to do all the casting.  I want to figure out a good way of exposing events, getting feedback from people trying to use it, and then make a nice event system for 0.10 and add this to the Client interface.

Note I had trouble parsing the event that came out - some json blob... maybe we can overhaul the json format for 0.10 as well...

ethanfrey commented:
Addressed some changes. Docs missing and thinking about how/if to expose websockets?  Maybe just expose an EventSwitch on the client? Does it make sense to work on that today?  Or rather targeting 0.9.1?

ebuchman commented on rpc/client/httpclient.go:
```diff
@@ -0,0 +1,349 @@
+	// used to maintain counts of actively listened events
+	// so we can properly subscribe/unsubscribe
+	// FIXME: thread-safety???
+	// FIXME: reuse code from go-events???
```
hmmm - why are we replicating all this code?

ethanfrey replied:
Because it is impossible to reuse code from go-event.  All private (package local).  I tried... then I did it.  As above, I am down to improve, but changes to go-event are out of scope.  Otherwise this PR will never get finished.

ethanfrey commented on rpc/client/helpers.go:
```diff
@@ -0,0 +1,89 @@
+// This handles subscribing and unsubscribing under the hood
+func WaitForOneEvent(evsw types.EventSwitch,
+	evtTyp string, timeout time.Duration) (types.TMEventData, error) {
+	listener := cmn.RandStr(12)
```
Which part is a bad idea?  Waiting?  subscribing for the events?

ebuchman replied:
sorry  I just meant using a random string for the listener. I'm just thinking of the day there might be many calls to WaitForOneEvent and no one will know which listener is which

ethanfrey replied:
the "listener" is just for that one call-back.  I can let the caller decide, but it will probably be hard-coded various places.  and duplicate names is worse, then second might not get the message.

ethanfrey commented on rpc/client/helpers.go:
```diff
@@ -0,0 +1,89 @@
+	evts, quit := make(chan events.EventData, 10), make(chan bool, 1)
+	// start timeout count-down
+	go func() {
+		time.Sleep(1 * time.Second)
```
huh?  comment on the name of the channel?
I have a timeout as time.Duration, so I picked another name.  Don't mind changing names, if you have an idea.

ebuchman replied:
shouldn't `time.Sleep(1 * time.Second)` be `time.Sleep(timeout * time.Second)`

ethanfrey commented on rpc/client/helpers.go:
```diff
@@ -0,0 +1,89 @@
+	})
+	// make sure to unregister after the test is over
+	// TODO: why doesn't the other call work???
+	// defer evsw.RemoveListenerForEvent(listener, evtTyp)
```
On client.Local, if I just call `RemoveListener`, it doesn't properly deregister.  And in the end the listener fills up the channel and the node blocks and tests hang.

It seemed to deep to debug (Node.EventSwitch()), so I just put a TODO.
In any case, the fix is out of scope of this pr.
